src/shared/lib/newsletter.ts

```diff
@@ -50,3 +54,21 @@ export const NewsletterMap = new Map<GeoLocation | undefined, Newsletters[]>([
     ] as Newsletters[],
   ],
 ]);
+
+// get a list of newsletters that have been updated in the body and compare
+// to list of all newsletter ids
+export const newslettersSubscriptionsFromFormBody = (body: {
+  [key: string]: unknown;
+}): NewsletterPatch[] =>
+  ALL_NEWSLETTER_IDS.map((id) => {
+    // if the id of a newsletter is included in the body
+    // then mark this newsletter as to potentially update (subscribe / unsubscribe)
+    // otherwise return undefined
+    if (id in body) {
+      return {
+        id,
+        subscribed: !!body[id],
+      };
+    }
+    // filter out the undefined values from the array to return the correct type
+  }).filter(Boolean) as NewsletterPatch[];
```

Comment:
You could use a simple forEach here and push to a separate array if id in body? Horses for courses and not a big deal or a blocker but it would be a way to avoid having to cast the type / filter out undefined?

I know the pattern of defining a new array and using forEach to populate it gets a bad rep but sometimes it does the job 🤷

Reply:
flatMap might be an option https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap#for_adding_and_removing_items_during_a_map

Reply:
updated to use `flatMap`, works perfectly 😁